File: ports/python/examples/screenshot.py
"""Renders a dashboard headlessly and prints it, so the whole stack can be
exercised without a TTY: ``python examples/screenshot.py``.

Add ``--ansi`` for the colored form, or ``--html`` for a standalone page.
"""

from __future__ import annotations

import math
import sys

import hqtui.widgets as w
from hqtui import Cell, GridSpec, Layout, Panel, render_to_html, render_to_screen
from hqtui.graphics import PlotOptions, Series


def main() -> None:
    arg = sys.argv[1] if len(sys.argv) > 1 else ""
    cpu = [math.sin(i / 9) * 35 + 55 for i in range(120)]
    net = [math.cos(i / 5) * 25 + 40 for i in range(120)]

    def view(root) -> None:
        def header(r) -> None:
            r.heading("hqtui — python port")
            r.spacer("fill")
            r.badge(w.BadgeOptions(text="LIVE"))

        def throughput(p) -> None:
            p.graph(
                w.GraphOptions(
                    series=[
                        Series(values=cpu, label="cpu"),
                        Series(values=net, label="net"),
                    ],
                    axis=True,
                    legend=True,
                    plot=PlotOptions(fill=True),
                )
            )

        def cores(p) -> None:
            p.meters(
                w.MetersOptions(
                    items=[w.MeterItem(f"c{i}", 0.15 + i * 0.11) for i in range(8)]
                )
            )

        def processes(p) -> None:
            p.table(
                w.TableOptions(
                    rows=[
                        w.TableRow(("1", "systemd", "0.1", "12M")),
                        w.TableRow(("412", "hqtui", "12.5", "48M")),
                        w.TableRow(("1201", "ruff", "41.8", "1.2G")),
                    ],
                    columns=[
                        w.TableColumn("PID", align="right"),
                        w.TableColumn("NAME"),
                        w.TableColumn("CPU%", align="right"),
                        w.TableColumn("MEM", align="right"),
                    ],
                    selected=1,
                    zebra=True,
                )
            )

        def grid(g) -> None:
            g.panel(Panel(title="Throughput", subtitle="60s"), Cell(), throughput)
            g.panel(Panel(title="Cores"), Cell(), cores)
            g.panel(Panel(title="Processes"), Cell(col_span=2), processes)

        root.row(Layout(size=1), header)
        root.grid(GridSpec(columns=["2fr", "1fr"], rows=[11, "1fr"], gap=1), grid)
        root.status_bar(
            w.StatusBarOptions(
                items=[w.StatusItem("quit", "q"), w.StatusItem("select", "↑↓")],
                right=[w.StatusItem("30fps")],
            )
        )

    screen = render_to_screen(84, 22, "dark", view)
    if arg == "--ansi":
        print(screen.ansi())
    elif arg == "--html":
        print(render_to_html(screen))
    else:
        print(screen.text())


if __name__ == "__main__":
    main()

File: ports/rust/examples/screenshot.rs
//! Renders a dashboard headlessly and prints it, so the whole stack can be
//! exercised without a TTY: `cargo run --example screenshot`.
//!
//! Add `--ansi` for the colored form, or `--html` for a standalone page.

use hqtui::graphics::Series;
use hqtui::prelude::*;
use hqtui::testing::{render_to_html, render_to_screen, HtmlOptions};

fn main() {
    let arg = std::env::args().nth(1).unwrap_or_default();
    let cpu: Vec<f64> = (0..120).map(|i| ((i as f64) / 9.0).sin() * 35.0 + 55.0).collect();
    let net: Vec<f64> = (0..120).map(|i| ((i as f64) / 5.0).cos() * 25.0 + 40.0).collect();

    let screen = render_to_screen(84, 22, "dark", move |ui| {
        ui.row(Row::new().size(1), |r| {
            r.heading("hqtui — rust port");
            r.spacer("fill");
            r.badge(BadgeOptions::new("LIVE"));
        });
        ui.grid(
            GridSpec::new()
                .columns(["2fr", "1fr"])
                .rows([Size::Cells(11), Size::Fr(1.0)])
                .gap(1),
            |g| {
                g.panel(
                    Panel::new().title("Throughput").subtitle("60s"),
                    Span::new(1, 1),
                    |p| {
                        p.graph(
                            GraphOptions::series(vec![
                                Series::new(cpu).label("cpu"),
                                Series::new(net).label("net"),
                            ])
                            .with_axis()
                            .with_legend()
                            .filled(),
                        );
                    },
                );
                g.panel(Panel::new().title("Cores"), Span::new(1, 1), |p| {
                    p.meters(MetersOptions {
                        items: (0..8)
                            .map(|i| MeterItem::new(format!("c{i}"), 0.15 + (i as f64) * 0.11))
                            .collect(),
                        ..Default::default()
                    });
                });
                g.panel(Panel::new().title("Processes"), Span::new(2, 1), |p| {
                    p.table(
                        TableOptions {
                            rows: vec![
                                TableRow::new(["1", "systemd", "0.1", "12M"]),
                                TableRow::new(["412", "hqtui", "12.5", "48M"]),
                                TableRow::new(["1201", "rustc", "41.8", "1.2G"]),
                            ],
                            columns: vec![
                                TableColumn::new("PID").align(Align::Right),
                                TableColumn::new("NAME"),
                                TableColumn::new("CPU%").align(Align::Right),
                                TableColumn::new("MEM").align(Align::Right),
                            ],
                            selected: Some(1),
                            zebra: true,
                            ..Default::default()
                        },
                        "",
                    );
                });
            },
        );
        ui.status_bar(StatusBarOptions {
            items: vec![StatusItem::new("quit").key("q"), StatusItem::new("select").key("↑↓")],
            right: vec![StatusItem::new("30fps")],
            ..Default::default()
        });
    });

    match arg.as_str() {
        "--ansi" => println!("{}", screen.ansi()),
        "--html" => println!("{}", render_to_html(&screen, &HtmlOptions::default())),
        _ => println!("{}", screen.text()),
    }
}
